api/agent/agent.go

```diff
@@ -807,6 +742,107 @@ func (a *agent) runHot(ctx context.Context, call *call, tok ResourceToken, state
 	logger.WithError(err).Info("hot function terminated")
 }
 
+// runHotReq enqueues a free slot to slot queue manager and watches various timers and the consumer until
+// the slot is consumed. A return value of false means, the container should shutdown and no subsequent
+// calls should be made to this function.
+func (a *agent) runHotReq(ctx context.Context, call *call, state ContainerState, logger logrus.FieldLogger, cookie drivers.Cookie, slot *hotSlot) bool {
+
+	var err error
+	isFrozen := false
+
+	freezeTimer := time.NewTimer(a.freezeIdleMsecs)
+	idleTimer := time.NewTimer(time.Duration(call.IdleTimeout) * time.Second)
+	ejectTicker := time.NewTicker(a.ejectIdleMsecs)
+
+	defer freezeTimer.Stop()
+	defer idleTimer.Stop()
+	defer ejectTicker.Stop()
+
+	// log if any error is encountered
+	defer func() {
+		if err != nil {
+			logger.WithError(err).Error("hot function failure")
+		}
+	}()
+
+	// if an immediate freeze is requested, freeze first before enqueuing at all.
+	if a.freezeIdleMsecs == time.Duration(0) && !isFrozen {
+		err = cookie.Freeze(ctx)
+		if err != nil {
+			return false
+		}
+		isFrozen = true
+	}
+
+	state.UpdateState(ctx, ContainerStateIdle, call.slots)
+	s := call.slots.queueSlot(slot)
+
+	for {
+		select {
+		case <-s.trigger: // slot already consumed, unfreeze ASAP
+			if !isFrozen {
```

Comment:
we have to freeze it just so that we can unfreeze it? seems a bit :(

Reply:
No, only if frozen. An optimization only, if s.trigger is triggered, we can proceed to unfreeze before attempting acquire. I'm going to remove this as it's hard to understand and perhaps premature.

Reply:
this code appears to do the opposite (it freezes if not frozen)

Reply:
arrghh, anyway, it's gone. :-)